src/supertux/autotile.cpp

```diff
@@ -0,0 +1,174 @@
+//  SuperTux
+//  Copyright (C) 2020 A. Semphris <semphris@protonmail.com>
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "supertux/autotile.hpp"
+
+#include "supertux/autotile_parser.hpp"
+
+// AutotileMask
+
+AutotileMask::AutotileMask(uint8_t mask, bool center) :
+  m_mask(mask),
+  m_center(center)
+{
+}
+
+bool
+AutotileMask::matches(uint8_t mask, bool center)
+{
+  return mask == m_mask && center == m_center;
+}
+
+// Autotile
+
+Autotile::Autotile(uint32_t tile_id, std::vector<AutotileMask*> masks, bool solid) :
+  m_tile_id(tile_id),
+  m_masks(masks),
+  m_solid(solid)
+{
+}
+
+bool
+Autotile::matches(uint8_t num_mask, bool center)
+{
+  for (auto& l_mask : m_masks)
+  {
+    if (l_mask->matches(num_mask, center))
+    {
+      return true;
+    }
+  }
+  return false;
+}
+
+uint32_t
+Autotile::get_tile_id()
+{
+  return m_tile_id;
+}
+
+bool
+Autotile::is_solid()
+{
+  return m_solid;
+}
+
+
+// AutotileSet
+
+std::vector<AutotileSet*>* AutotileSet::m_autotilesets = new std::vector<AutotileSet*>();
+
+AutotileSet::AutotileSet(std::vector<Autotile*> tiles, uint32_t default_tile) :
+  m_autotiles(tiles),
+  m_default(default_tile)
+{
+}
+
+AutotileSet*
+AutotileSet::get_tileset_from_tile(uint32_t tile_id)
+{
```

Comment:
I'm not sure if this method can be marked `const` as well.

Reply:
That one isn't only a read; it's also the one that populates m_autotilesets if it is empty.